File: Data Structures & Algorithms/max-water-container/submission-3.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l = len(heights)

        if l == 0:
            return 0

        i = 0
        j = l - 1
        maxSquare = 0

        while i < j:
            if heights[i] > heights[j]:
                newSquare = (j - i) * heights[j]

                j -= 1
            else:
                newSquare = (j - i) * heights[i]

                i += 1

            if maxSquare < newSquare:
                maxSquare = newSquare    

        return maxSquare

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-3.java
class Solution {
    private class Item {
        private final int height;
        private final int width;

        private Item(int height, int width) {
            this.height = height;
            this.width = width;
        }
    }

    public int largestRectangleArea(int[] heights) {
        Stack<Item> stack = new Stack<Item>();
        int maxArea = 0;

        for (int i = 0; i <= heights.length; i++) {
            int h = 0;
            if (i < heights.length) {
                h = heights[i];
            }

            int width = 1;
            for (;;) {
                if (stack.empty()) {
                    stack.push(new Item(h, width));

                    break;
                } else {
                    Item top = stack.peek();

                    if (top.height >= h) {
                        stack.pop();

                        width += top.width;
                        int area = (width - 1) * top.height;
                        if (area > maxArea) {
                            maxArea = area;
                        }
                    } else {
                        stack.push(new Item(h, width));

                        break;
                    }
                }
            }
        }

        return maxArea;
    }
}